client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -682,21 +682,54 @@ private void setHeaders(HttpRequest httpRequest, Collection<Header> requestHeade
          * node.
          */
         if (false == deadNodes.isEmpty()) {
-            Collections.sort(deadNodes);
-
-            List<Node> selectedDeadNodes = new ArrayList<>(deadNodes.size());
-            for (DeadNodeAndRevival n : deadNodes) {
-                selectedDeadNodes.add(n.node);
-            }
-            nodeSelector.select(selectedDeadNodes);
+            final List<DeadNodeAndRevival> selectedDeadNodes = new ArrayList<>(deadNodes);
+            /*
+             * We'd like NodeSelectors to remove items directly from deadNodes
+             * so we can find the minimum after it is filtered without having
+             * to compare many things. This saves us a sort on the unfiltered
+             * list.
+             */
+            nodeSelector.select(new Iterable<Node>() {
+                @Override
+                public Iterator<Node> iterator() {
+                    return new Adapter(selectedDeadNodes.iterator());
+                }
+            });
             if (false == selectedDeadNodes.isEmpty()) {
-                return singletonList(selectedDeadNodes.get(0));
+                return singletonList(Collections.min(selectedDeadNodes).node);
```

Comment:
maybe silly question: how does `min` help compared to sorting? Doesn't min iterate over all the elements?